pkg/reconciler/pubsub/topic/topic.go

```diff
@@ -0,0 +1,441 @@
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logging.FromContext(ctx).Desugar().Error("Invalid resource key")
+		return nil
+	}
+
+	// Get the Topic resource with this namespace/name
+	original, err := r.topicLister.Topics(namespace).Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Desugar().Error("Topic in work queue no longer exists")
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	topic := original.DeepCopy()
+
+	// Reconcile this copy of the Topic and then write back any status
+	// updates regardless of whether the reconciliation errored out.
+	var reconcileErr = r.reconcile(ctx, topic)
+
+	// If no error is returned, mark the observed generation.
+	if reconcileErr == nil {
+		topic.Status.ObservedGeneration = topic.Generation
+	}
+
+	if equality.Semantic.DeepEqual(original.Finalizers, topic.Finalizers) {
+		// If we didn't change finalizers then don't call updateFinalizers.
+
+	} else if _, updated, fErr := r.updateFinalizers(ctx, topic); fErr != nil {
+		logging.FromContext(ctx).Desugar().Warn("Failed to update Topic finalizers", zap.Error(fErr))
+		r.Recorder.Eventf(topic, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update finalizers for Topic %q: %v", topic.Name, fErr)
+		return fErr
+	} else if updated {
+		// There was a difference and updateFinalizers said it updated and did not return an error.
+		r.Recorder.Eventf(topic, corev1.EventTypeNormal, "Updated", "Updated Topic %q finalizers", topic.Name)
+	}
+
+	if equality.Semantic.DeepEqual(original.Status, topic.Status) {
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+
+	} else if _, uErr := r.updateStatus(ctx, topic); uErr != nil {
+		logging.FromContext(ctx).Desugar().Warn("Failed to update Topic status", zap.Error(uErr))
+		r.Recorder.Eventf(topic, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update status for Topic %q: %v", topic.Name, uErr)
+		return uErr
+	} else if reconcileErr == nil {
+		// There was a difference and updateStatus did not return an error.
+		r.Recorder.Eventf(topic, corev1.EventTypeNormal, "Updated", "Updated Topic %q", topic.Name)
+	}
+	if reconcileErr != nil {
+		r.Recorder.Event(topic, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
+	}
+	return reconcileErr
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, topic *v1alpha1.Topic) error {
+	ctx = logging.WithLogger(ctx, r.Logger.With(zap.Any("topic", topic)))
+
+	topic.Status.InitializeConditions()
+
+	if topic.DeletionTimestamp != nil {
+		logging.FromContext(ctx).Desugar().Debug("Deleting Pub/Sub topic")
```

Comment:
Nit: Only log this if the topic will actually be deleted. Move the log statement inside the if block below.

Reply:
Done!